designs/toggling-provisioners.md

```diff
@@ -0,0 +1,183 @@
+
+### Maintenance Scope/API Surface
+
+#### In-line with Provisioner
+
+In-lining Maintenance with a Provisioner sets Maintenance Windows to refer to only nodes spawned by that Provisioner. For users that use a Provisioner per type of workload, or per groupings of applications, an in-line Provisioner Maintenance Window takes advantage of nodes being owned by a Provisioner.
+
+```
+spec:
+    maintenanceCron: "* * * * 1-5"
+    maintenanceDurationSeconds: 3600
+```
+
+The inherent downside of an in-line Provisioner is the lack of extensibility to/across Provisioners and across pods. If a cluster operator wants to stop deprovisioning across the whole cluster, they need to copy this Maintenance Window to every provisioner in the cluster to ensure no nodes are deprovisioned. In addition, if a cluster operator’s Provisioner requirements are highly flexible, being more driven at the pod-level, executing Maintenance Windows at a Provisioner level may inhibit users from taking advantage of Karpenter’s provisioning magic while enabling application-based maintenance.
+
+#### Custom Resource
+
+Implementing a [Custom Resource](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) means that Karpenter will have to support API Versions and vending more custom resources through the Helm Chart for installation. Creating a dedicated CR for Maintenance Windows easens placement of additional features in the future.
+
+#### Exclusionary Windows
+
+Users with clusters generally tolerant to disruption could use Exclusionary Windows to represent their busiest hours, specifying actions that cannot be used.
+
+An Exclusionary Window is the inverse of a Maintenance Window and could be implemented as a boolean value in a Maintenance Window or as a separate Exclusionary Window CR. *We’ll refer to a Maintenance Window as a Deny Window, and an Exclusionary Window as an Allow Window.*
+
+*Alternative 1:*
+Merge both types of windows into one Custom Resource.
+
+```
+apiVersion: karpenter.sh/v1alpha1
+kind: MaintenanceWindow
+metadata:
+  name: disable-everything
+spec:
+  // no selector, matches everything
+  // no schedules, matches everything
+  type: Allow | Deny
+  actions:
+    - Expiration // disable expiration
+    - Consolidation
+    - Drift
+    - Rebalance
+```
+
+*Alternative 2:*
+Keep both types of windows as two separate Custom Resources. Since they would have identical specs, this redundancy would be better put into a type field as above.
+
+```
+apiVersion: karpenter.sh/v1alpha1
+kind: MaintenanceWindow
+metadata:
+  name: disable-everything
+spec:
+  // no selector, matches everything
+  // no schedules, matches everything
+  actions:
+    - Expiration // disable expiration
+    - Consolidation
+    - Drift
+    - Rebalance
+---
+apiVersion: karpenter.sh/v1alpha1
+kind: ExclusionaryWindow
```

Comment:
I like the idea of calling a Maintenance Window as exclusionary someway as it will exclude actions that Karpenter can take, but I worry about MaintenanceExclusionPeriod being too long. What are your thoughts?